File: leetcode/LC/design/0359_rate_limitor.py
class Logger:
    
    def __init__(self):
        self.init_time = 0
        self.prev = {}
        self.curr = {}

    def shouldPrintMessage(self, timestamp: int, message: str) -> bool:
        if(timestamp >= self.init_time + 10):
            self.prev = self.curr.copy()
            self.curr.clear()
            self.init_time = timestamp
        if(self.curr.get(message) != None):
            return False
        if(self.prev.get(message) and self.prev[message] + 10 > timestamp):
            return False
        self.curr[message] = timestamp
        return True

File: leetcode/LC1/design/0359_rate_limitor.cpp

// //https://leetcode.com/problems/logger-rate-limiter/discuss/365306/Simple-Two-HashMap-Solution-with-O(1)-time-and-little-memory/640609
#include <unordered_map>
#include <string>
using namespace std;

class Logger {
public:
    unordered_map <string, int> prev;
    unordered_map <string, int> curr;
    int init_time;
    Logger() {    
        init_time = 0;
    }
    
    bool shouldPrintMessage(int timestamp, string message) {
        if(timestamp >= init_time + 10){
            prev = curr;
            curr.clear();
            init_time = timestamp;
        }
        
        if(curr.find(message) != curr.end())
           return false;
        if(prev.find(message) != prev.end() && prev[message] + 10 > timestamp)
           return false;
        curr[message] = timestamp;
        return true;
    }
};
